stacks/api/bucket/src/graphql/graphql.ts

```diff
@@ -0,0 +1,609 @@
+        bucket._id,
+        {_id: documentId},
+        input,
+        {
+          returnOriginal: true
+        }
+      );
+
+      const currentDocument = {...input, _id: documentId};
+
+      if (this.activity) {
+        const _ = this.insertActivity(context, Action.PUT, bucket._id, documentId);
+      }
+
+      if (this.history) {
+        const promise = createHistory(
+          this.bs,
+          this.history,
+          bucket._id,
+          previousDocument,
+          currentDocument
+        );
+      }
+
+      let aggregation = [];
+
+      let requestedFields = requestedFieldsFromInfo(info);
+      let relationAndLocalization = await aggregationsFromRequestedFields(
+        JSON.parse(JSON.stringify(bucket)),
+        requestedFields,
+        this.getLocale,
+        JSON.parse(JSON.stringify(this.buckets))
+      );
+      aggregation.push(...relationAndLocalization);
+
+      aggregation.push({$match: {_id: documentId}});
+
+      let project = getProjectAggregation(requestedFields);
+      aggregation.push(project);
+
+      return this.bds.find(bucket._id, aggregation).then(([documents]) => documents);
+    };
+  }
+
+  patch(bucket: Bucket): Function {
+    return async (
+      root: any,
+      {_id: documentId, input}: {[arg: string]: any},
+      context: any,
+      info: GraphQLResolveInfo
+    ) => {
+      await this.authenticate(
+        context,
+        "/bucket/:bucketId/data/:documentId",
+        {bucketId: bucket._id, documentId: documentId},
+        ["bucket:data:update"],
+        {resourceFilter: false}
+      );
+
+      let previousDocument = await this.bds.findOne(bucket._id, {_id: documentId});
+
+      let patchedDocument = getPatchedDocument(previousDocument, input);
+
+      await this.validateInput(bucket._id, patchedDocument).catch(error =>
+        throwError(error.message, 400)
+      );
+
+      let updateQuery = getUpdateQuery(previousDocument, patchedDocument);
+
+      if (!Object.keys(updateQuery).length) {
+        throw Error("There is no difference between previous and current documents.");
+      }
+
+      let currentDocument = await this.bds.findOneAndUpdate(
+        bucket._id,
+        {_id: documentId},
+        updateQuery,
+        {returnOriginal: false}
+      );
+
+      if (this.history) {
+        const promise = createHistory(
+          this.bs,
+          this.history,
+          bucket._id,
+          previousDocument,
+          currentDocument
+        );
+      }
+
+      if (this.activity) {
+        const _ = this.insertActivity(context, Action.PUT, bucket._id, documentId);
+      }
+
+      let aggregation = [];
+
+      let requestedFields = requestedFieldsFromInfo(info);
+      let relationAndLocalization = await aggregationsFromRequestedFields(
+        JSON.parse(JSON.stringify(bucket)),
+        requestedFields,
+        this.getLocale,
+        JSON.parse(JSON.stringify(this.buckets))
```

Comment:
Fetching all these buckets upfront and (un)serializing them is intense work.  
We should fetch buckets on an as-needed basis to lift off the pressure from the database and memory.

Reply:
This has been done in several places. We have to fix them as well.